btree/btree.c

```diff
@@ -1501,6 +1502,253 @@ static void generic_move(struct nd *src, struct nd *tgt,
+					  struct m0_be_tx_credit *accum)
+{
+	struct m0_be_tx_credit kv_update_cred;
+
+	kv_update_cred = M0_BE_TX_CREDIT(1, ksize + vsize);
+	btree_mem_alloc_credit(tree, ksize + vsize, accum);
+	m0_be_tx_credit_add(accum, &kv_update_cred);
+}
+
+static void btree_insert_leaf_node_credit(const struct m0_btree  *tree,
+					  m0_bcount_t             ksize,
+					  m0_bcount_t             vsize,
+					  struct m0_be_tx_credit *accum)
+{
+	m0_btree_create_node_credit(tree, 1, accum);
+	btree_move_key_value_credit(accum, MAX_MOVE);
+	btree_insert_key_value_credit(tree, ksize, vsize, accum);
+}
+
+static void btree_insert_internal_node_credit(const struct m0_btree  *tree,
+					      struct m0_be_tx_credit *accum)
+{
+	m0_btree_create_node_credit(tree, 1, accum);
+	btree_move_key_value_credit(accum, MAX_MOVE);
+	btree_insert_key_value_credit(tree, max_ksize, internal_vsize, accum);
+}
+
+static void btree_insert_new_root_credit(const struct m0_btree  *tree,
+					 struct m0_be_tx_credit *accum)
+{
+	m0_btree_create_node_credit(tree, 1, accum);
+	btree_move_key_value_credit(accum, ALL_MOVE);
+
+	/*credit for inserting 2 key-value at root node*/
+	struct m0_be_tx_credit cred = {};
+	btree_insert_key_value_credit(tree, max_ksize, internal_vsize, &cred);
+	m0_be_tx_credit_mac(accum, &cred, 2);	
+}
+
+/**
+ * Calculates total credits required to perform @nr insert operations
+ * @param nr     Number of insert operations.
+ * @param ksize  Key data size.
+ * @param vsize  Value data size.
+*/
+static void insert_credit(const struct m0_btree  *tree,
+			  m0_bcount_t             nr,
+			  m0_bcount_t             ksize,
+			  m0_bcount_t             vsize,
+			  struct m0_be_tx_credit *accum)
+{
+	uint32_t height = BTREE_HEIGHT_MAX;
+	
+	btree_insert_leaf_node_credit(tree, ksize, vsize, accum);
+
+	struct m0_be_tx_credit cred = {};
+	btree_insert_internal_node_credit(tree, &cred);
+	m0_be_tx_credit_mac(accum, &cred, height - 1);
+
+	btree_insert_new_root_credit(tree, accum);
+
+	m0_be_tx_credit_mul(accum, nr);
+}
+
+static void btree_mem_free_credit(const struct m0_btree  *btree,
+				  m0_bcount_t             size,
+				  struct m0_be_tx_credit *accum)
+{
+	m0_be_allocator_credit(NULL, M0_BAO_FREE_ALIGNED, 0, 0, accum);
+}
+
+/**
+ * Calculates the credit needed to free @nr nodes and adds this credit
+ * to @accum.
+ */
+static void m0_btree_free_node_credit(const struct m0_btree  *tree,
+				      m0_bcount_t             nr,
+				      struct m0_be_tx_credit *accum)
+{
+	struct m0_be_tx_credit cred = {};
+
+	btree_mem_free_credit(tree, max_node_size, &cred);
+	m0_be_tx_credit_mac(accum, &cred, nr);
+}
+
+static void btree_delete_key_value_credit(const struct m0_btree  *tree,
+					  m0_bcount_t             ksize,
+					  m0_bcount_t             vsize,
+					  struct m0_be_tx_credit *accum)
+{
+	btree_mem_free_credit(tree, ksize + vsize, accum);
+	m0_be_tx_credit_add(accum,
+			    &M0_BE_TX_CREDIT_TYPE(struct m0_btree_rec));
```

Comment:
Deleting key-value may trigger node merge?

Reply:
I have consider credits required for  merging btree_delete_leaf_node_credit() and btree_delete_internal_node_credit(), more specifically, btree_move_key_value_credit(accum, MIN_MOVE); provide credits for merging/ rebalancing two nodes.